transport/http2_client.go

```diff
@@ -554,17 +569,32 @@ func (t *http2Client) handleSettings(f *http2.SettingsFrame) {
 	}
 	f.ForeachSetting(func(s http2.Setting) error {
 		if v, ok := f.Value(s.ID); ok {
-			t.mu.Lock()
-			defer t.mu.Unlock()
 			switch s.ID {
 			case http2.SettingMaxConcurrentStreams:
-				t.maxStreams = v
+				// TODO(zhaoq): This is a hack to avoid significant refactoring of the
+				// code to deal with the unrealistic int32 overflow. Probably will try
+				// to find a better way to handle this later.
+				if v > math.MaxInt32 {
+					v = math.MaxInt32
+				}
+				t.mu.Lock()
+				reset := t.streamsQuota != nil
+				ms := t.maxStreams
+				t.maxStreams = int(v)
+				t.mu.Unlock()
+				if !reset {
+					t.streamsQuota = newQuotaPool(int(v))
+				} else {
+					t.streamsQuota.reset(int(v) - ms)
```

Comment:
A minimal example:  what if you have two newStream calls that both update maxStreams, from X to Y and then to Z?

The first call does:

    Lock
    ms := X
    t.maxStreams = Y
    Unlock

The second does:

    Lock
    ms := Y
    t.maxStreams = Z
    Unlock
    reset(Z - Y)

And then back to the first:

    reset(Y - X)

And that's not even taking into account the t.streamsQuota.add calls that may have occurred in the interim, which AFAICT introduce problems of their own.

(In general, when you're mixing atomic-increment and atomic-reset the resets need to be compare-and-swap, not blind writes.)